scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/TaskFlowJob.java

```diff
@@ -170,11 +165,41 @@ public Set<Task> findTerminalTasks() {
             if (elseChildrenTaskNames.contains(task.getName())) {
                 terminalTasks.removeAll(findSubTree(task));
             }
+
         }
         // all remaining tasks are terminal
         return terminalTasks;
     }
 
+    /**
+     * Find root tasks for this workflow
+     *
+     * Root tasks are tasks without parent and not targets of a If, Else or Continuation branch
+     * @return a set of terminal tasks
+     */
+    public Set<Task> findRootTasks() {
+
+        Set<Task> rootTasks = new LinkedHashSet<>(tasks.values());
+
+        for (Task task : tasks.values()) {
+
+            FlowScript flowScript = task.getFlowScript();
+            if (flowScript != null && flowScript.getActionType().equals(FlowActionType.IF.toString())) {
+                if (flowScript.getActionTarget() != null) {
+                    // remove from root all tasks that are target of this task's IF, ELSE or CONTINUATION branches
+                    rootTasks.removeIf(t -> t.getName().equals(flowScript.getActionTarget()));
+                    rootTasks.removeIf(t -> t.getName().equals(flowScript.getActionTargetElse()));
+                    rootTasks.removeIf(t -> t.getName().equals(flowScript.getActionContinuation()));
```

Comment:
You can do it in a single `removeIf()`, this will avoid to potentially call two more uncessary removeIf. Also you can read the removeIf result and if true add a continue to bypass second if.

Reply:
right

Reply:
though both if are independant, I will do the first change you mention

Reply:
Yes you are right. In any case it would not have been a valuable change as you would still have to test the `removeIf ` result.